### PR TITLE
Maintain access to genParticles from signal event in packed collection for HI miniAOD

### PR DESCRIPTION
**Pull request by mandrenguyen**
#### PR description:

For event-overlay MC workflows, we need to know which particles are from the signal event, i.e., the one with collisionID =0.  This method is not available in the packedGenParticle collection.  This PR adds a PackedGenParticleRefVector containing the PackedGenParticles with collisionID =0.  It's run for all HI eras; there is no change to pp workflows. 

#### PR validation:

Tested with 158.01

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Will need an 11_2_X backport

@ttrk @stepobr @stahlleiton

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
assign xpog
(perhaps more appropriate than reco for `packedGenParticles`)

**Comment by perrotta** [comment]
DataFormats/PatCandidates and PhysicsTools/PatAlgos were just removed from the Analysis category: I wonder whether they should stay under xpog now (in addition to reconstruction): @cms-sw/xpog-l2 what do you think?

**Comment by mandrenguyen** [comment]
Do I interpret correctly that the failing test has to do with input files?

**Comment by slava77** [comment]
> Do I interpret correctly that the failing test has to do with input files?

no; the workflows outputs are less easy to find now

in the results "Matrix INPUT Tests Outputs" https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-56c8c2/12284/runTheMatrixINPUT-results/140.5611_RunHI2018AOD+RunHI2018AOD+REMINIAODHID18+HARVESTHI18MINIAOD/step2_RunHI2018AOD+RunHI2018AOD+REMINIAODHID18+HARVESTHI18MINIAOD.log
```
An exception of category 'ProductNotFound' occurred while
   [0] Processing  Event run: 326479 lumi: 15 event: 3808176 stream: 0
   [1] Running path 'MINIAODoutput_step'
   [2] Prefetching for module PoolOutputModule/'MINIAODoutput'
   [3] Calling method for module PackedGenParticleSignalProducer/'packedGenParticlesSignal'
Exception Message:
Principal::getByToken: Found zero products matching all criteria
Looking for type: edm::Association<std::vector<pat::PackedGenParticle> >
Looking for module label: packedGenParticles
Looking for productInstanceName: 
```

**Comment by slava77** [comment]
genParticles should not be running on data

**Comment by slava77** [comment]
>  the workflows outputs are less easy to find now

that was a wrong statement; I corrected it now. We have more short tests done and they are placed in the "INPUTRelVals" category of tests.

**Comment by perrotta** [comment]
please test

File: PhysicsTools/PatAlgos/plugins/PackedGenParticleSignalProducer.cc [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+void pat::PackedGenParticleSignalProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  const auto& genParticles = iEvent.getHandle(genParticleToken_);
+  const auto& orig2packed = iEvent.get(assoToken_);
+
+  auto signalGenParticleRefs = std::make_unique<pat::PackedGenParticleRefVector>();
+
+  for (auto it = genParticles->begin(); it != genParticles->end(); ++it) {
+    const auto& orig = reco::GenParticleRef(genParticles, it - genParticles->begin());
+    const auto& packed = orig2packed[orig];
```
Comment (perrotta) [inline PhysicsTools/PatAlgos/plugins/PackedGenParticleSignalProducer.cc:41]: This should be moved after the check `if (orig->collisionId() != 0)`

File: PhysicsTools/PatAlgos/plugins/PackedGenParticleSignalProducer.cc [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+#include "DataFormats/PatCandidates/interface/PackedGenParticle.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+
+namespace pat {
+
+  class PackedGenParticleSignalProducer : public edm::stream::EDProducer<> {
```
Comment (perrotta) [inline PhysicsTools/PatAlgos/plugins/PackedGenParticleSignalProducer.cc:12]: This could also become a `global` module, I guess

**Comment by perrotta** [comment]
In 100 events of wf 158.01, 2.3 kB are taken by these  PackedGenParticlesRefs, some 2.8% of the overall miniAOD size:

```
Compare packed values
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
      0.0 ->      2300.4       2300     NEWO   2.84     patPackedGenParticlesRefs_packedGenParticlesSignal__DQM.
-------------------------------------------------------------
    81000 ->       83300       2301             2.8     ALL BRANCHES
```

Size comparison with other GenParticles in the same events:
```
packedGenParticles: (patPackedGenParticles) 330417 34336.3
prunedGenParticles: (recoGenParticles) 25347.2 5424.39
packedGenParticlesSignal: (patPackedGenParticlesRefs) 17059 2300.39
 ```

**Comment by mandrenguyen** [comment]
@perrotta The MC input file in relval is not representative.  It's a pythia event embedded into a very, very peripheral (b=12) PbPb event.  Running on normal embedded MC, the output miniAOD is only a fraction of a percent smaller. 
The file I tested is:
/store/himc/HINPbPbAutumn18DR/Bjet_pThat-15_TuneCP5_HydjetDrumMB_5p02TeV_Pythia8/AODSIM/mva98_103X_upgrade2018_realistic_HI_v11-v1/30000/7D8C962E-9A46-9541-94AA-69C01C817F2E.root

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- The new code adds a PackedGenParticleRefVector with refs to selected "signal" generated particles, in HI MC
- Jenkins tests pass and show no difference but the additional collection in HI MC workflows

**Comment by mariadalfonso** [comment]
+xpog

**Comment by qliphy** [comment]
+1

**Pull request merged**